components/GlobalMenu/CategoriesMenu.js

```diff
@@ -6,92 +6,148 @@
  * See LICENSE
  */
 
-import React, { Fragment } from 'react';
+import React, { Fragment, type Node } from 'react';
 import { Trans } from '@lingui/react';
 
-import type { Categories } from './index';
+import type { ReadingLevel } from '../../types';
+import { fetchCategories } from '../../fetch';
 import Link from '../BrowseLink';
 import ReadingLevelTrans from '../ReadingLevelTrans';
 import Menu, { MenuItem } from '../Menu';
+import { ActivityIndicator } from '../../elements';
+import { spacing } from '../../style/theme';
+
+type Categories = {
+  classroom_books?: Array<ReadingLevel>,
+  library_books?: Array<ReadingLevel>
+};
 
 type Props = {|
-  categories: Categories,
   languageCode: string,
-  onClose: (
-    event:
-      | SyntheticMouseEvent<any>
-      | SyntheticKeyboardEvent<any>
-      | KeyboardEvent
-  ) => void
+  // Render prop
+  children: (data: { onClick: () => void }) => Node,
+  onSelectCategory: () => void,
+  openStateCallback: boolean => void
 |};
 
-export default class CategoriesMenu extends React.Component<Props> {
+type State = {
+  showMenu: boolean,
+  categories: ?Categories
+};
+
+export default class CategoriesMenu extends React.Component<Props, State> {
+  state = {
+    categories: null,
+    showMenu: false
+  };
+
+  handleShowMenu = () => {
+    if (!this.state.languages) {
+      this.loadCategories();
+    }
+
+    this.props.openStateCallback(true);
+    this.setState({ showMenu: true });
+  };
+
+  handleCloseMenu = () => {
+    this.props.openStateCallback(false);
+    this.setState({ showMenu: false });
+  };
+
+  async loadCategories() {
+    const categoriesRes = await fetchCategories(this.props.languageCode);
+
+    // TODO: Handle error case by notifying user?
+    if (categoriesRes.isOk) {
```

Comment:
Kva skjer dersom kallet ikkje er ok?

Reply:
Evig spinner for brukeren

Reply:
Opprettet https://github.com/GlobalDigitalLibraryio/issues/issues/327